File: frontend/src/services/userCourseService.js
import apiClient, {
  API_USER_COURSES,
  API_USER_COURSES_COUNT,
  API_USER_COURSES_BY_USER,
  API_USER_COURSES_BY_COURSE,
  API_USER_COURSES_COUNT_BY_USER,
  API_USER_COURSES_COUNT_BY_COURSE,
  API_USER_COURSE_BY_ID,
} from './api';

class UserCourseService {
  // Create a new user-course relationship
  async createUserCourse(userCourseData) {
    try {
      console.log('UserCourseService: Creating user-course with data:', userCourseData);
      const response = await apiClient.post(API_USER_COURSES, userCourseData);
      console.log('UserCourseService: Create response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error creating user-course relationship:', error);
      console.error('Error response:', error.response?.data);
      throw error;
    }
  }

  // Get all user-course relationships with pagination
  async getUserCourses(skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_USER_COURSES, {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationships:', error);
      throw error;
    }
  }

  // Get user-course relationship by ID
  async getUserCourseById(userCourseId) {
    try {
      const response = await apiClient.get(API_USER_COURSE_BY_ID(userCourseId));
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationship:', error);
      throw error;
    }
  }

  // Update user-course relationship
  async updateUserCourse(userCourseId, userCourseData) {
    try {
      console.log('UserCourseService: Updating user-course', userCourseId, 'with data:', userCourseData);
      const response = await apiClient.put(API_USER_COURSE_BY_ID(userCourseId), userCourseData);
      console.log('UserCourseService: Update response:', response.data);
      return response.data;
    } catch (error) {
      console.error('Error updating user-course relationship:', error);
      console.error('Error response:', error.response?.data);
      throw error;
    }
  }

  // Delete user-course relationship
  async deleteUserCourse(userCourseId) {
    try {
      const response = await apiClient.delete(API_USER_COURSE_BY_ID(userCourseId));
      return response.data;
    } catch (error) {
      console.error('Error deleting user-course relationship:', error);
      throw error;
    }
  }

  // Get user-course relationships count
  async getUserCoursesCount() {
    try {
      const response = await apiClient.get(API_USER_COURSES_COUNT);
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationships count:', error);
      throw error;
    }
  }

  // Get user-course relationships by user
  async getUserCoursesByUser(userId, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_USER_COURSES_BY_USER(userId), {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationships by user:', error);
      throw error;
    }
  }

  // Get user-course relationships by course
  async getUserCoursesByCourse(courseId, skip = 0, limit = 10) {
    try {
      const response = await apiClient.get(API_USER_COURSES_BY_COURSE(courseId), {
        params: { skip, limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationships by course:', error);
      throw error;
    }
  }

  // Get user-course relationships count by user
  async getUserCoursesCountByUser(userId) {
    try {
      const response = await apiClient.get(API_USER_COURSES_COUNT_BY_USER(userId));
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationships count by user:', error);
      throw error;
    }
  }

  // Get user-course relationships count by course
  async getUserCoursesCountByCourse(courseId) {
    try {
      const response = await apiClient.get(API_USER_COURSES_COUNT_BY_COURSE(courseId));
      return response.data;
    } catch (error) {
      console.error('Error fetching user-course relationships count by course:', error);
      throw error;
    }
  }
}

export default new UserCourseService(); 

File: backend/src/services/user_course_service.py
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from src.model.user_course import UserCourse, UserCourseCreate, UserCourseResponse
from typing import List, Optional
import logging

logger = logging.getLogger(__name__)

class UserCourseService:
    """Service class for handling user-course relationship business logic."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_user_course(self, user_course_data: UserCourseCreate) -> UserCourseResponse:
        try:
            db_user_course = UserCourse(**user_course_data.model_dump())
            self.db.add(db_user_course)
            self.db.commit()
            self.db.refresh(db_user_course)
            logger.info(f"Created user-course relationship with ID: {db_user_course.id}")
            return UserCourseResponse.model_validate(db_user_course)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to create user-course relationship due to integrity error: {e}")
            raise ValueError("User-course relationship creation failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to create user-course relationship: {e}")
            raise ValueError(f"User-course relationship creation failed: {str(e)}")
    
    def get_user_course_by_id(self, user_course_id: int) -> Optional[UserCourseResponse]:
        user_course = self.db.query(UserCourse).filter(UserCourse.id == user_course_id).first()
        if user_course:
            return UserCourseResponse.model_validate(user_course)
        return None
    
    def get_user_courses_by_user(self, user_id: int, skip: int = 0, limit: int = 10) -> List[UserCourseResponse]:
        user_courses = self.db.query(UserCourse).filter(
            UserCourse.user_id == user_id
        ).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(user_courses)} user-course relationships for user {user_id}")
        return [UserCourseResponse.model_validate(uc) for uc in user_courses]
    
    def get_user_courses_by_course(self, course_id: int, skip: int = 0, limit: int = 10) -> List[UserCourseResponse]:
        user_courses = self.db.query(UserCourse).filter(
            UserCourse.course_id == course_id
        ).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(user_courses)} user-course relationships for course {course_id}")
        return [UserCourseResponse.model_validate(uc) for uc in user_courses]
    
    def list_user_courses(self, skip: int = 0, limit: int = 10) -> List[UserCourseResponse]:
        user_courses = self.db.query(UserCourse).offset(skip).limit(limit).all()
        logger.info(f"Retrieved {len(user_courses)} user-course relationships with skip={skip}, limit={limit}")
        return [UserCourseResponse.model_validate(uc) for uc in user_courses]
    
    def update_user_course(self, user_course_id: int, user_course_data: UserCourseCreate) -> Optional[UserCourseResponse]:
        db_user_course = self.db.query(UserCourse).filter(UserCourse.id == user_course_id).first()
        if not db_user_course:
            return None
        try:
            for key, value in user_course_data.model_dump().items():
                setattr(db_user_course, key, value)
            self.db.commit()
            self.db.refresh(db_user_course)
            logger.info(f"Updated user-course relationship with ID: {db_user_course.id}")
            return UserCourseResponse.model_validate(db_user_course)
        except IntegrityError as e:
            self.db.rollback()
            logger.error(f"Failed to update user-course relationship {user_course_id} due to integrity error: {e}")
            raise ValueError("Update failed: integrity constraint violation")
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to update user-course relationship {user_course_id}: {e}")
            raise ValueError(f"Update failed: {str(e)}")
    
    def delete_user_course(self, user_course_id: int) -> bool:
        db_user_course = self.db.query(UserCourse).filter(UserCourse.id == user_course_id).first()
        if not db_user_course:
            return False
        try:
            self.db.delete(db_user_course)
            self.db.commit()
            logger.info(f"Deleted user-course relationship with ID: {user_course_id}")
            return True
        except Exception as e:
            self.db.rollback()
            logger.error(f"Failed to delete user-course relationship {user_course_id}: {e}")
            raise ValueError(f"Deletion failed: {str(e)}")
    
    def get_user_courses_count(self) -> int:
        return self.db.query(UserCourse).count()
    
    def get_user_courses_count_by_user(self, user_id: int) -> int:
        return self.db.query(UserCourse).filter(UserCourse.user_id == user_id).count()
    
    def get_user_courses_count_by_course(self, course_id: int) -> int:
        return self.db.query(UserCourse).filter(UserCourse.course_id == course_id).count() 